File: Data Structures & Algorithms/linked-list-cycle-detection/submission-1.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        visited = set()

        def dfs(node):
            if not node: return False

            if node.val in visited:
                return True
                
            visited.add(node.val)

            return dfs(node.next)

        return dfs(head)

File: Data Structures & Algorithms/remove-node-from-end-of-linked-list/submission-1.js
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     constructor(val = 0, next = null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */

class Solution {
    /**
     * @param {ListNode} head
     * @param {number} n
     * @return {ListNode}
     */
    removeNthFromEnd(head, n) {
        const dummy = new ListNode(0, head)
        let left = dummy
        let right = head
        let activeIdx = 0
        while (activeIdx < n) {
            right = right.next
            activeIdx += 1
        }
        while (right){
            right = right.next
            left = left.next
        }
        left.next = left.next.next
        return dummy.next
        
    }
}
